cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java

```diff
@@ -130,50 +119,14 @@ private TriggeringScheduleInfo getTriggeringScheduleInfo(Job job) {
       if (nsPrincipal != null && SecurityUtil.isKerberosEnabled(cConf)) {
         SecurityRequestContext.setUserId(new KerberosName(nsPrincipal).getServiceName());
       }
-      runtimeInfo = lifecycleService.startInternal(id, sysArgs, userArgs, false);
+      ProgramOptions programOptions = new SimpleProgramOptions(id, new BasicArguments(sysArgs),
+                                                               new BasicArguments(userArgs), false);
+      provisionerNotifier.provisioning(id.run(RunIds.generate()), programOptions);
```

Comment:
Why interact with `ProvisionerNotifier` directly? I would expect the AppLifecycleService is the only integration point for other components.

Reply:
The reason for that is because there would be new lifecycle methods to be run before provision and after deprovisioning. Exposing the internal mechanism outside of the AppLifecycleService breaks the abstraction boundary.

Reply:
I was running into issues starting the ProgramLifecycleService in unit test because there is a cyclic dependency for guice injection with ProgramLifecycleService -> CoreSchedulerService -> ConstraintCheckerService -> ProgramLifecycleService. Though it must be working somehow so I'll take another look at the guice modules and see what can be done.

Reply:
Seems like the circular dependency is a wrong construct. The `ProgramLifecycleService` should take a `Scheduler`, since those functions (e.g. getScheduleStatus, suspendResumeSchedule, etc.) belongs to the Scheduler service, not program lifecycle.

Reply:
agreed, I may open a separate PR to do this refactoring first